x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/Fork.java

```diff
@@ -85,4 +132,40 @@ public boolean equals(Object o) {
 
         return Objects.equals(children(), other.children());
     }
+
+    @Override
+    public BiConsumer<LogicalPlan, Failures> postAnalysisPlanVerification() {
+        return Fork::checkFork;
+    }
+
+    private static void checkFork(LogicalPlan plan, Failures failures) {
+        if (plan instanceof Fork == false) {
+            return;
+        }
+        Fork fork = (Fork) plan;
+
+        Map<String, DataType> outputTypes = fork.children()
+            .getFirst()
+            .output()
+            .stream()
+            .collect(Collectors.toMap(Attribute::name, Attribute::dataType));
+
+        fork.children().stream().skip(1).forEach(subPlan -> {
+            for (Attribute attr : subPlan.output()) {
+                var actual = attr.dataType();
+                var expected = outputTypes.get(attr.name());
+                if (actual != expected) {
```

Comment:
If I understand correctly, then there are no widening conversations, e.g. integer to long. E.g.
```
| FORK ( EVAL a = 1 )
       ( EVAL a = 20000000000000000 )
```

I think that this is fine, just wanted to call it out explicitly. Maybe worth adding a test, and something to consider for the future (as a potential follow-up).

Reply:
got a test already here:

https://github.com/elastic/elasticsearch/blob/abce76200fd1ef0d402aa6bcdd1d0e91512c800a/x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/ForkIT.java#L562-L572

added as a follow up in https://github.com/elastic/elasticsearch/issues/121652